llvm/lib/Object/ArchiveWriter.cpp

```diff
@@ -482,16 +482,47 @@ static uint64_t computeHeadersSize(object::Archive::Kind Kind,
 }
 
 static Expected<std::unique_ptr<SymbolicFile>>
-getSymbolicFile(MemoryBufferRef Buf, LLVMContext &Context) {
+getSymbolicFile(MemoryBufferRef Buf, LLVMContext &Context,
+                object::Archive::Kind Kind) {
   const file_magic Type = identify_magic(Buf.getBuffer());
   // Don't attempt to read non-symbolic file types.
   if (!object::SymbolicFile::isSymbolicFile(Type, &Context))
     return nullptr;
   if (Type == file_magic::bitcode) {
     auto ObjOrErr = object::SymbolicFile::createSymbolicFile(
         Buf, file_magic::bitcode, &Context);
-    if (!ObjOrErr)
-      return ObjOrErr.takeError();
+    // An error reading a bitcode file most likely indicates that the file
+    // was created by a compiler from the future. Normally we don't try to
+    // implement forwards compatibility for bitcode files, but when creating an
+    // archive we can implement best-effort forwards compatibility by treating
+    // the file as a blob and not creating symbol index entries for it. lld and
+    // mold ignore the archive symbol index, so provided that you use one of
+    // these linkers, LTO will work as long as lld or the gold plugin is newer
+    // than the compiler. We only ignore errors if the archive format is one
+    // that is supported by a linker that is known to ignore the index,
+    // otherwise there's no chance of this working so we may as well error out.
+    // We print a warning on read failure so that users of linkers that rely on
+    // the symbol index can diagnose the issue.
+    //
+    // This is the same behavior as GNU ar when the linker plugin returns an
+    // error when reading the input file. If the bitcode file is actually
+    // malformed, it will be diagnosed at link time.
+    if (!ObjOrErr) {
+      switch (Kind) {
+      case object::Archive::K_BSD:
+      case object::Archive::K_GNU:
+      case object::Archive::K_GNU64:
+        llvm::logAllUnhandledErrors(ObjOrErr.takeError(), llvm::errs(),
```

Comment:
I'm not keen on this warning being printed inside the library. This will prevent e.g. client tools treating the warning as an error. I would prefer one of two options:

1) Continue returning it up the stack, and report the warning further up (perhaps using something in the error code within the error to identify it as opposed to other genuine errors).

2) Pass in a callback function that is called when this case is hit. Client code can choose then to handle the error in whatever way it chooses.

Reply:
We don't really have a system for warnings with different warning levels though, at least not outside of Clang. Taking your example of a client that wants warnings to be errors, it would not be sufficient to return it up the stack from here because `Expected` is a sum type not a product type. Such a client would likely not want the file to be created at all if there were a warning, so the callback may be a better choice to support such a client (which would do something like return true if the passed-in error should actually be treated as an error) but to some degree that's just an elaborate way of passing in a `bool WarningsAreErrors` argument, and maybe that's all that the client would want or need.

So I think we should avoid trying to design for hypothetical clients and instead add features on an as-needed basis, especially if the way to support the hypothetical client is not obvious. All in-tree clients want a warning printed to stderr, so that's what I implemented here.

Reply:
I agree with the analysis. `llvm::logAllUnhandledErrors` is probably not the most elegant approach, but it is the most practical one. Passing in `bool WarningsAreErrors` would require updating quite a few functions and end up adding lots of complexity.

`writeArchive` is primarily used by in-tree clients. A hypothetical client, even if exists, would unlikely create an archive with a new bitcode member that would cause the warning.

Reply:
>  but to some degree that's just an elaborate way of passing in a bool WarningsAreErrors argument, and maybe that's all that the client would want or need.

I disagree here - the important difference is that, not just the choice fo whether to continue, but the choice of where/how to render the warning is in the hands of the client - this is core to LLVM's API-centric design. If an API user is multithreaded and wants to buffer output, or render it differently (in a GUI), etc, they should be able to.

> So I think we should avoid trying to design for hypothetical clients and instead add features on an as-needed basis, especially if the way to support the hypothetical client is not obvious. All in-tree clients want a warning printed to stderr, so that's what I implemented here.

I don't think we need to design for hypothetical clients - but hold to a fairly generic goal that LLVM's is designed as reusable library components and it's not too speculative/arbitrary to not print directly from a library because clients may have different needs for how their output is produced/rendered/handled.

Reply:
> I don't think we need to design for hypothetical clients - but hold to a fairly generic goal that LLVM's is designed as reusable library components and it's not too speculative/arbitrary to not print directly from a library because clients may have different needs for how their output is produced/rendered/handled.

I hear what you're saying but if we say "don't write to stderr from a library" then we need to answer "what should we do instead". If LLVM already had a diagnostic API with programmatic control over diagnostic levels and  `writeArchive` were already using it then it would be a no-brainer to expose the warning via that API. But if that isn't actually implemented yet then at that point it becomes non-obvious what the right thing to do is, and we need to do *something* -- and that's when we start needing to think about what the client might need (i.e. "design for hypothetical clients"). If/when the actual client comes along that needs this we'd probably discover that what we designed was not sufficient and would need to redo the API for the client's needs, which would take more effort overall than just designing it for the actual client in the first place. And if a client doesn't care about any of this and just wants us to write to stderr they would need to change their usage code multiple times for no benefit to them.

I suppose the minimal thing we could do here is add a `llvm::raw_ostream &Warn` to `writeArchive` and that's where the warnings would go. A client that wants warnings as errors can error out and delete the output file if anything was written to the stream, a buffered client can save the output and write it in one go and a GUI can split on newlines. That seems fairly minimal, doesn't require writing dead code and should be easy enough to replace with something else if it becomes necessary. But really I think we should just wait for an actual client.

Reply:
In other places (I'm thinking specifically of the DWARF debug line parser), we use a callback system, where a `std::function<void(Error &&)>`-like functor is passed into the function that might produce the warning. The error is passed to the callback, and the caller does what it wants with it. I believe we have a default value for the callback somewhere in the callstack to just print to stderr, which means that only clients who care have to actually modify their signatures.

This approach provides complete flexibility to the client, so we don't have to consider further what clients would want to do, whilst also providing an easy route for the known existing clients (e.g. lld) who just want the warnings to be printed to stderr.

Reply:
So you think the code should look like this?
```
static Expected<std::unique_ptr<SymbolicFile>>
getSymbolicFile(MemoryBufferRef Buf, LLVMContext &Context,
                object::Archive::Kind Kind, function_ref<void(Error)> Warn) {
[...]
      case object::Archive::K_BSD:
      case object::Archive::K_GNU:
      case object::Archive::K_GNU64:
        Warn(ObjOrErr.takeError());
        return nullptr;
```
It seems basically the same as `raw_ostream &Warn`. It will only easily accommodate clients that want to write the warning somewhere else and will require clients wanting warnings as errors to delete the file afterwards. So I think `raw_ostream &Warn` is better (but doing nothing is best), because it makes the client code simpler in the 99+% case.

Reply:
Adding a parameter requires updates to quite a few functions including `writeArchive` and `writeArchiveToBuffer`. `writeArchive` has a few in-tree callers that need adjustment. I personally don't want to do this.

If @dwblaikie and @jh7370  are strong about this, I think `function_ref<...> = []() { ... }` with a default argument is actually the easiest since in-tree callers do not need adjustment.

Reply:
> If @dwblaikie and @jh7370 are strong about this, I think `function_ref<...> = []() { ... }` with a default argument is actually the easiest since in-tree callers do not need adjustment.

That will work well but only as long as that's the last argument. If we add another argument X to `writeArchive` and a caller wants to pass a non-default value for X they'll need to copy-paste the lambda out of the declaration. With `std::raw_ostream &Warn` they'll only need to copy-paste `llvm::errs()`.

This sort of thing is why I'm generally against expanding APIs without a concrete use case. It might seem like a simple change to add a parameter here but we could end up with unintended consequences as a result of future API evolution that adds to the maintenance burden for ourselves and other clients.

Reply:
reusability is a cornerstone of LLVM (best citation I have for that is the first sentence on llvm.org: "The LLVM Project is a collection of modular and reusable compiler and toolchain technologies."), and that means not using global state (including global streams like stdout/stderr) and not exiting (returning errors instead) - admittedly we have some violations (eg: report_fatal_error, lld) but it's a generally held to convention.

I'd like to see something like https://reviews.llvm.org/D65515 - though I suspect a good handler might be one that takes an Error by rvalue, and returns an Error by value (to allow the warning handler to say "just treat this as an error and  don't continue"). Though I don't feel so strongly about that particular variation.

Please add a callback that may have a default (that prints to stderr or the like) - but that /can/ be specified via llvm-objdump.cpp for instance (eg: it's not much use if there's a default argument to a function call, but then there are unconditional calls to that function withuot specifying the handler that are still internal to libObject).

I guess libObject doesn't have any sort of context object this could be readily attached to? None that I could find, at a glance?

Reply:
I looked at implementing your suggestion. Here's the patch I came up with on top of this one:
https://github.com/pcc/llvm-project/commit/llvm-ar-warn-example

I'm sorry, but I don't think it's a good API. So I still want to keep things as is here until a client needs something else.

After implementing it I discovered the following problems:
1. Warnings are of poor quality. For example, without the added call to `createFileError` and with the changes to llvm-ar.cpp reverted, the warning looks like this:
```
warning: Invalid bitcode signature
```
This is because of the error wrapping in the caller, which does not occur when calling the `Warn` function directly. I partially mitigated it by adding the call to `createFileError`. This leads to problem 2.
2. The change does not properly support the warnings-as-errors use case. In the change, I made a change to llvm-ar.cpp to simulate a hypothetical client that wants warnings as errors. After the change, we can see an ugly error message occur:
```
ra/bin/llvm-ar: error: bad2.a: 'input.bc': 'ra/obj/llvm/test/Object/Output/archive-malformed-object.test.tmp.dir/input.bc': Invalid bitcode signature
```
Note that `input.bc` is repeated. In order to fix this, more extensive refactoring is needed to ensure that `createFileError` is only called once. We'll also need to pass in the correct file argument wherever an error is created. I don't think this extensive refactoring should be done just so that we can print one warning. Instead, it should be the responsibility of the future contributor who wants this behavior in their client. Of course, if you want to do the refactoring yourself, you are welcome to do so in a followup.

Reply:
I dig some digging into our downstream LLVM usage, and we actually have a downstream patch to LLD for a downstream-only piece of functionality that ultimately results in a potential call to `writeArchive` (with symbol table enabled) from within LLD, with an archive that may contain bitcode files. I believe with this PR the flow goes something like this:
1) The downstream feature is fed one or more invalid bitcode files and attempts to create an archive from them via `writeArchive`.
2) With this PR, the call will result in a warning, due to the invalid bitcode file, printed directly to stderr, ignoring LLD's usual warning mechanism (which includes `--fatal-warnings`).
3) The downstream feature eventually results in the archive being compiled for LTO, which presumably will result in an error due to the unrecognised bitcode file, meaning we ultimately end up with, in theory, a warning and an error.

In this concrete use-case, we either don't want the warning, or we'd want it to be an error, since it's indicative of a future issue and we should stop earlier if possible, but even if we did want just a warning, it really should end up going through LLD's warning mechanism.

If we've got a downstream case where the warning shouldn't be written directly to stderr, I fully expect there to be other such cases.

Reply:
My earlier message highlighted a problem with the `llvm::Error` API: it doesn't really handle "contexts" very well, such as the file name. If we really wanted to handle this "properly", I think we would need to fundamentally redesign `llvm::Error` to handle contexts for all diagnostics, not just errors. For example, we could make it so that `llvm::Expected` holds a list of warnings on the side and those get propagated through callers when adding contexts. And then we wouldn't want the callback at all, we could just handle it in the caller. An alternative approach would involve a completely different way of denoting contexts, such as by using RAII objects.

In your use case, I think the `raw_ostream &Warn` argument that I described before is sufficient. You can redirect the output stream to a buffer and use it to detect whether a warning was printed, and then print the warnings to stderr followed by e.g. "error: warnings treated as errors". LLD's warning API just prints to stderr so it'll be more or less the same except that it won't be colored. That's a minor aesthetic detail which I think we can put up with for now until the `llvm::Error` API is redesigned.

Reply:
@bd1976bris, could you comment on the above and whether it works for our downstream use-case, please?

I personally don't like it, as I don't think a stream passed in is any better than passing in the suggested callback.

Reply:
Callback based error handling with context seems plausible - by wrapping when context is added. Eg, if some previous codepath only produced errors, it might look like this:
```
Expected<std::unique_ptr<SymbolicFile>> SymFileOrErr =
    getSymbolicFile(M.Buf->getMemBufferRef(), Context);
if (!SymFileOrErr)
  return createFileError(M.MemberName, SymFileOrErr.takeError());
```
The buggy/missing context version would be to just pass in the warning callback here (from the client code):
```
Expected<std::unique_ptr<SymbolicFile>> SymFileOrErr =
    getSymbolicFile(M.Buf->getMemBufferRef(), Context, WarnHandler);
```
but then that could be improved to provide context with a wrapper, first by having the handler take an Error, not a string, then that would allow the call site to wrap for additional context:
```
Expected<std::unique_ptr<SymbolicFile>> SymFileOrErr =
    getSymbolicFile(M.Buf->getMemBufferRef(), Context, [&](Error E) {
    return WarnHandler(createFileError(M.MemberName, std::move(E)));
});
```
Which could be simplified for reuse with a utility:
```
Expected<std::unique_ptr<SymbolicFile>> SymFileOrErr =
    getSymbolicFile(M.Buf->getMemBufferRef(), Context, wrapWarningWithFile(WarnHandler));
```
Or similar.

Having clients have to trawl through text to find the "warning: " text embedded in the original text, or somehow have to add it inline in the stream after each newline, or something else seems pretty messy compared to this ^.

Reply:
Ah, I think I'm following - if a warning becomes an error via the callback, then it goes back up the stack as an error and gets context added again. (doesn't quite explain the /different/ context in the example - one with the full path name, one with the short name... )

Fair question. I wouldn't be averse to omitting the pass-through behavior on the first pass of this - and having warnings-as-errors be implemented entirely in the caller (they see a warning, emit it as an error (so I think the callback should still pass an Error) but provide no return value from the handler to allow it to be transformed into an error - caller can then record they emitted an error and combine that with any error result/failure code coming from the main API)

If someone wanted to enhance this to support the early-exit-on-werror behavior, they could have the API wrap and then unwrap the Error if it was returned from the handler.

Reply:
> I personally don't like it, as I don't think a stream passed in is any better than passing in the suggested callback.

I'm sorry that you don't like it. I don't like it either (I don't think we'll find an API that all of us like, this seems like a very bikesheddable topic), but it seems like a reasonable compromise to me. The reason for the stream based API is that we know that `Error` is the wrong API for warning use cases because of its poor handling of contexts, so we should avoid propagating it into the caller. The stream is a minimal API extension that'll be easier to change later.

> Having clients have to trawl through text to find the "warning: " text embedded in the original text, or somehow have to add it inline in the stream after each newline, or something else seems pretty messy compared to this ^.

That's not what a client would need to do. As I said, the messages added to the buffer will already contain "warning: ", so they will just print the buffer text as-is followed by "error: warnings treated as errors". That's not ideal but it's good enough, and if it doesn't work for someone they are welcome to propose their own patch with a better error handling API.

> doesn't quite explain the /different/ context in the example - one with the full path name, one with the short name... 

That's `Buf.getBufferIdentifier()` (full name) vs `M.MemberName` (short name). The latter isn't passed into `getSymbolicFile` but I suppose it could be passed in as well to make the warning a bit more consistent.

> I wouldn't be averse to omitting the pass-through behavior on the first pass of this - and having warnings-as-errors be implemented entirely in the caller (they see a warning, emit it as an error (so I think the callback should still pass an Error) but provide no return value from the handler to allow it to be transformed into an error - caller can then record they emitted an error and combine that with any error result/failure code coming from the main API)

So when `getSymbolicFile` returns a warning it will do
```
Warn(createFileError(MemberName, ObjOrErr.takeError());
return nullptr;
```
and when it returns an error it will do
```
return ObjOrErr.takeError();
```
?

I'm not sure about that. If we only add the context sometimes it will read like a mistake. Another approach would be to remove the call to `createFileError` from the caller of `getSymbolicFile` and add calls to `createFileError` to `getSymbolicFile` whenever it creates an error or warning. So now warnings and errors look the same:
```
return errorOrValue(Warn(createFileError(MemberName, ObjOrErr.takeError())), nullptr); // warning
```
or
```
return createFileError(MemberName, ObjOrErr.takeError()); // error
```
But that's generally more error prone, it'll be easier to miss adding a call to `createFileError` especially if we follow the pattern elsewhere in the code to return warnings.

Reply:
Could we have each layer of the callstack do the minimal context addition necessary so that context isn't lost? In this particular case, the getSymbolicFile function wouldn't need to do anything other than pass the error to the callback directly, the caller would add the member name via a wrapper around the callback, since it is the function that knows it, and then this process would proceed up the stack as more additional context is gained at each level (obviously cases in the stack that don't have any additional context to add don't need to wrap the callback themselves). We do something similar in the Object library in some places, e.g. [here](https://github.com/llvm/llvm-project/blob/93869dfd89387844bf8b605ebcd1abc0cc81bde8/llvm/lib/Object/ELFObjectFile.cpp#L899): in this case the lower code reports the low-level error, the code in the link adds information about the referencing section, and the calling code will ultimately add the file name, before it all ends up getting printed as a warning in llvm-readobj.

I acknowledge the situation is slightly different here, since we're distinguishing between warnings and errors, whereas llvm-readobj treats most things as warnings, so can continue to pass Error instances up the stack, but the overall approach should still apply, just using callbacks (with each layer wrapping the callback if needed), or even a string passed back upwards that is added to then ultimately passed to a callback.

Reply:
> > doesn't quite explain the /different/ context in the example - one with the full path name, one with the short name...
> 
> That's `Buf.getBufferIdentifier()` (full name) vs `M.MemberName` (short name). The latter isn't passed into `getSymbolicFile` but I suppose it could be passed in as well to make the warning a bit more consistent.

I was thinking the opposite direction -if the current error handling doesn't have its context passed in, but attached later when going up the callstack, the analogous behavior with an warning callback would be to attach the context in the callback (at whatever layer was adding it to the error, we'd wrap the callback in a "context adding callback" as discussed in a few places in this thread)

But it does present issues if that warning can then be passed back out of the callback and become an error - which then gets the context added to it again.

I would be OK with not addressing that issue for now - having a callback of `void(Error)`, doing the context wrapping as described ^ (so it has the same context as the error used to have, added in the same level in the call stack).

I think it's reasonable to leave the "if I want a werror-like behavior that also early-exits, I should figure out how to deal with context adding during warning without then duplicating that context addition when the warning-promoted-to-error gets propagated up the stack too" - I can think of a few possible solutions to that problem, but I think they're worth having as a separate discussion closer to the use case.

Reply:
I had the same idea about wrapping the lambda but I'm not a fan of the amount of boilerplate needed when calling a function that might warn.
```
return createFileName(FileName, foo([&](Error Err) { Warn(createFileName(FileName, Err)); });
```
And you still need a post-facto fixup in the caller -- except now it's "was a callback called" and not "is a buffer empty". So I think this makes the code worse than `raw_ostream &Warn` because it's more subtle, more verbose in the callee and probably the caller, less intuitive, and shoehorns bad APIs into places where they don't fit, making it harder to remove them. But I guess it's not that much worse, it only happens in one place for now so I guess I'm not entirely opposed to doing that.

Side note, here is the usage code for an API that I would consider replacing `Error` with:
```
void foo(StringRef FileName, DiagHandler &Diag) {
  auto FDiag = Diag.withFileName(FileName);
  bar(FDiag);
  if (FDiag.hasErrors())
    return;
  baz(FDiag);
}

void bar(DiagHandler &Diag) {
   Diag.warn("some warning"); // could be promoted to error by Diag implementation
   Diag.error("some error");
}
```

Reply:
Done